# Homework

In this homework, we'll deploy the Straight vs Curly Hair Type model we trained in the previous homework.

Download the model from here:

https://github.com/alexeygrigorev/large-datasets/releases/download/hairstyle/model_2024_hairstyle.keras

In [1]:
!wget https://github.com/alexeygrigorev/large-datasets/releases/download/hairstyle/model_2024_hairstyle.keras

--2024-12-11 18:47:37--  https://github.com/alexeygrigorev/large-datasets/releases/download/hairstyle/model_2024_hairstyle.keras
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/426348925/df5735c1-9082-4b67-968e-866f268793f8?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241211%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241211T114607Z&X-Amz-Expires=300&X-Amz-Signature=7e49844b3dc6f35fa4e63779ca7b57dc2521a28f834481a388b5fd5d1909e926&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dmodel_2024_hairstyle.keras&response-content-type=application%2Foctet-stream [following]
--2024-12-11 18:47:38--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/426348925/df5735c1-9082-4b67-968e-866f268793f8?X-

In [2]:
import tensorflow as tf
import os

# Load the Keras model
keras_model = tf.keras.models.load_model("model_2024_hairstyle.keras")

# Convert the model to TF-Lite format
converter = tf.lite.TFLiteConverter.from_keras_model(keras_model)
tflite_model = converter.convert()

# Save the TF-Lite model
tflite_model_path = "model_2024_hairstyle.tflite"
with open(tflite_model_path, "wb") as f:
    f.write(tflite_model)

print(f"Model successfully converted to TF-Lite: {tflite_model_path}")

# Check the size of the converted model
size_in_kb = os.path.getsize(tflite_model_path) / 1024 /1024
print(f"Size of the TF-Lite model: {size_in_kb:.2f} MB")

2024-12-11 18:48:05.406359: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-12-11 18:48:05.409856: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:32] Could not find cuda drivers on your machine, GPU will not be used.
2024-12-11 18:48:05.446308: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:32] Could not find cuda drivers on your machine, GPU will not be used.
2024-12-11 18:48:05.507345: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1733917685.546954  137872 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1733917685.55

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/garjita/.local/lib/python3.10/site-packages/ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "/home/garjita/.local/lib/python3.10/site-packages/traitlets/config/application.py", line 1075, in launch_instance
    app.start()
  File "/home/garjita/.local/lib/python3.10/site-

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/garjita/.local/lib/python3.10/site-packages/ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "/home/garjita/.local/lib/python3.10/site-packages/traitlets/config/application.py", line 1075, in launch_instance
    app.start()
  File "/home/garjita/.local/lib/python3.10/site-

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/garjita/.local/lib/python3.10/site-packages/ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "/home/garjita/.local/lib/python3.10/site-packages/traitlets/config/application.py", line 1075, in launch_instance
    app.start()
  File "/home/garjita/.local/lib/python3.10/site-

AttributeError: _ARRAY_API not found

W0000 00:00:1733917689.564342  137872 gpu_device.cc:2344] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...


INFO:tensorflow:Assets written to: /tmp/tmps6n_azq3/assets


INFO:tensorflow:Assets written to: /tmp/tmps6n_azq3/assets


Saved artifact at '/tmp/tmps6n_azq3'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): TensorSpec(shape=(None, 200, 200, 3), dtype=tf.float32, name='input_layer')
Output Type:
  TensorSpec(shape=(None, 1), dtype=tf.float32, name=None)
Captures:
  140145883180448: TensorSpec(shape=(), dtype=tf.resource, name=None)
  140145883218160: TensorSpec(shape=(), dtype=tf.resource, name=None)
  140145883595168: TensorSpec(shape=(), dtype=tf.resource, name=None)
  140145883597280: TensorSpec(shape=(), dtype=tf.resource, name=None)
  140145883605552: TensorSpec(shape=(), dtype=tf.resource, name=None)
  140145883602384: TensorSpec(shape=(), dtype=tf.resource, name=None)


W0000 00:00:1733917690.284965  137872 tf_tfl_flatbuffer_helpers.cc:365] Ignored output_format.
W0000 00:00:1733917690.285062  137872 tf_tfl_flatbuffer_helpers.cc:368] Ignored drop_control_dependency.
2024-12-11 18:48:10.285610: I tensorflow/cc/saved_model/reader.cc:83] Reading SavedModel from: /tmp/tmps6n_azq3
2024-12-11 18:48:10.285899: I tensorflow/cc/saved_model/reader.cc:52] Reading meta graph with tags { serve }
2024-12-11 18:48:10.285904: I tensorflow/cc/saved_model/reader.cc:147] Reading SavedModel debug info (if present) from: /tmp/tmps6n_azq3
I0000 00:00:1733917690.288506  137872 mlir_graph_optimization_pass.cc:401] MLIR V1 optimization pass is not enabled
2024-12-11 18:48:10.290804: I tensorflow/cc/saved_model/loader.cc:236] Restoring SavedModel bundle.
2024-12-11 18:48:10.353783: I tensorflow/cc/saved_model/loader.cc:220] Running initialization op on SavedModel bundle at path: /tmp/tmps6n_azq3
2024-12-11 18:48:10.358701: I tensorflow/cc/saved_model/loader.cc:466] SavedModel 

Model successfully converted to TF-Lite: model_2024_hairstyle.tflite
Size of the TF-Lite model: 76.58 MB


## Question 1
Now convert this model from Keras to TF-Lite format.

What's the size of the converted model?

- 27 Mb
- 43 Mb
- 77 Mb
- 127 Mb

### Answer 1: 77 Mb

## Question 2
To be able to use this model, we need to know the index of the input and the index of the output.

What's the output index for this model?

- 3
- 7
- 13
- 24

In [3]:
import tensorflow as tf

# Load the TF-Lite model
tflite_model_path = "model_2024_hairstyle.tflite"
interpreter = tf.lite.Interpreter(model_path=tflite_model_path)

# Allocate tensors
interpreter.allocate_tensors()

# Get input and output tensor details
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Print input details
print("Input details:")
for i, detail in enumerate(input_details):
    print(f"Input Index {i}: {detail}")

# Print output details
print("\nOutput details:")
for i, detail in enumerate(output_details):
    print(f"Output Index {i}: {detail}")

# Print the output index directly
output_index = output_details[0]['index']
print(f"\nThe output index for this model is: {output_index}")


Input details:
Input Index 0: {'name': 'serving_default_input_layer:0', 'index': 0, 'shape': array([  1, 200, 200,   3], dtype=int32), 'shape_signature': array([ -1, 200, 200,   3], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0), 'quantization_parameters': {'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}, 'sparsity_parameters': {}}

Output details:
Output Index 0: {'name': 'StatefulPartitionedCall_1:0', 'index': 13, 'shape': array([1, 1], dtype=int32), 'shape_signature': array([-1,  1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0), 'quantization_parameters': {'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}, 'sparsity_parameters': {}}

The output index for this model is: 13


INFO: Created TensorFlow Lite XNNPACK delegate for CPU.


### Answer 2: 13

## Preparing the image

You'll need some code for downloading and resizing images. You can use this code:

In [4]:
from io import BytesIO
from urllib import request

from PIL import Image

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img


def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img

In [5]:
# For that, you'll need to have pillow installed:
!pip install pillow

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: /usr/bin/python3 -m pip install --upgrade pip


Let's download and resize this image:

https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg

Based on the previous homework, what should be the target size for the image?

In [ ]:
# !wget https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg

## Question 3
Now we need to turn the image into numpy array and pre-process it.

Tip: Check the previous homework. What was the pre-processing we did there?

After the pre-processing, what's the value in the first pixel, the R channel?

- 0.24
- 0.44
- 0.64
- 0.84

In [6]:
import numpy as np
from io import BytesIO
from urllib import request
from PIL import Image

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img

def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img

# Target size for the model
target_size = (224, 224)

# Download and prepare the image
url = "https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg"
image = download_image(url)
image = prepare_image(image, target_size)

# Convert image to numpy array and pre-process it
image_array = np.array(image) / 255.0  # Scale pixel values between 0 and 1

# Check the first pixel value in the R channel
r_channel_first_pixel = image_array[0, 0, 0]  # Top-left pixel, R channel
print(f"Value in the first pixel, R channel: {r_channel_first_pixel:.7f}")


Value in the first pixel, R channel: 0.2392157


## Answer 3: 0.24

## Question 4
Now let's apply this model to this image. What's the output of the model?

- 0.293
- 0.493
- 0.693
- 0.893

In [7]:
import tensorflow as tf
import numpy as np
from PIL import Image
from urllib import request
from io import BytesIO

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img

def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img

# Download and prepare the image
url = "https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg"
target_size = (200, 200)
image = download_image(url)
image = prepare_image(image, target_size)

# Convert to NumPy array and preprocess
image_array = np.array(image) / 255.0  # Scale to [0, 1]
input_data = np.expand_dims(image_array, axis=0).astype(np.float32)  # Shape: (1, 200, 200, 3)

# Load the TF-Lite model
interpreter = tf.lite.Interpreter(model_path="model_2024_hairstyle.tflite")
interpreter.allocate_tensors()

# Get input and output details
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Set the input tensor
interpreter.set_tensor(input_details[0]['index'], input_data)

# Run inference
interpreter.invoke()

# Get the output tensor
output_data = interpreter.get_tensor(output_details[0]['index'])
print(f"Output of the model: {output_data[0][0]:.3f}")


Output of the model: 0.894


### Answer 4: 0.893

### Prepare the lambda code
Now you need to copy all the code into a separate python file. You will need to use this file for the next two questions.

Tip: you can test this file locally with ipython or Jupyter Notebook by importing the file and invoking the function from this file.

### Docker
For the next two questions, we'll use a Docker image that we already prepared. This is the Dockerfile that we used for creating the image:

```
FROM public.ecr.aws/lambda/python:3.10
COPY model_2024_hairstyle_v2.tflite .
RUN pip install numpy==1.23.1
```

Note that it uses Python 3.10. The latest models of TF Lite do not support Python 3.12 yet, so we need to use 3.10. Also, for this part, we will use TensorFlow 2.14.0. We have tested it, and the models created with 2.17 could be served with 2.14.0.

For that image, we also needed to use an older version of numpy (1.23.1)

The docker image is published to [agrigorev/model-2024-hairstyle:v3](https://hub.docker.com/r/agrigorev/model-2024-hairstyle/tags)

A few notes:

- The image already contains a model and it's not the same model as the one we used for questions 1-4.
- The wheel for this combination that you'll need to use in your Docker image is https://github.com/alexeygrigorev/tflite-aws-lambda/raw/main/tflite/tflite_runtime-2.14.0-cp310-cp310-linux_x86_64.whl

In [8]:
!docker pull agrigorev/model-2024-hairstyle:v3

v3: Pulling from agrigorev/model-2024-hairstyle
Digest: sha256:6d0aa19127467401fd439c76e662ceba2bd21a5713f174d4f89b354a26939ea6
Status: Downloaded newer image for agrigorev/model-2024-hairstyle:v3
docker.io/agrigorev/model-2024-hairstyle:v3


In [9]:
!docker images |grep agrigorev/model-2024-hairstyle

agrigorev/model-2024-hairstyle   v3           607850a2138e   7 days ago      782MB


## Question 5
Download the base image agrigorev/model-2024-hairstyle:v3. You can do it with docker pull.

So what's the size of this base image?

- 182 Mb
- 382 Mb
- 582 Mb
- 782 Mb

You can get this information when running docker images - it'll be in the "SIZE" column.

### Answer 5: 782 Mb

## Question 6
Now let's extend this docker image, install all the required libraries and add the code for lambda.

You don't need to include the model in the image. It's already included. The name of the file with the model is model_2024_hairstyle_v2.tflite and it's in the current workdir in the image (see the Dockerfile above for the reference). The provided model requires the same preprocessing for images regarding target size and rescaling the value range than used in homework 8.

Now run the container locally.

Score this image: https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg

What's the output from the model?

- 0.229
- 0.429
- 0.629
- 0.829

## Build image (execute in command line)
docker build -t hairstyle-predictor-lambda .

## Run the Container (execute in command line)
docker run -p 9000:8080 hairstyle-predictor-lambda


## Invoke Lambda Function (Testing with JSON)
Note: Install jp 
```
sudo apt-get install jq
```

In [11]:
!curl -X POST "http://localhost:9000/2015-03-31/functions/function/invocations" -d '{"image_url": "https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg"}' | jq .

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   139  100    51  100    88     34     59  0:00:01  0:00:01 --:--:--    93
{
  "statusCode": 200,
  "body": {
    "prediction": 0.9131
  }
}


### Answer 6: 0.829